### PR TITLE
[do-not-review] [cudagraph] Capture pipeline forward-backward steps

### PR DESCRIPTION
**Pull request by xmfan**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* (to be filled)

## Why

TorchTitan captures the non-pipeline forward-backward function, but pipeline training still executes every schedule eagerly. The pipeline schedule is a suitable graph boundary once input preprocessing remains outside capture and communication order, process groups, tensor structure, and tensor metadata remain static.

This change applies the same full-forward-backward capture model to pipeline schedules.

## Capture boundary

```text
Eager code
    |
    +--> load microbatches
    +--> move tensors to the device
    +--> preprocess model inputs
    |
    v
CUDA graph boundary
    |
    +--> pipeline schedule
    +--> forward computation
    +--> pipeline P2P communication
    +--> loss computation
    +--> backward computation
    +--> detached microbatch-loss reduction
    |
    v
Eager code
    |
    +--> gradient clipping
    +--> optimizer and scheduler steps
    +--> metrics reduction and logging
```

## Change

Split pipeline preprocessing from `_pp_forward_backward_body()` and wrap that schedule body when CUDA graphs are enabled. Use a persistent graph-owned loss sentinel on ranks that do not own the last stage.

Expose the upstream deferred-receive control for looped schedules and configure directed P2P process groups before distributed initialization. CUDA-graph pipeline training requires either explicit per-direction process groups or PyTorch automatic TorchComms groups.

Reject pipeline CUDA graphs with validation because validation reinitializes the shared schedule and invalidates the captured training graph.

## Testing

The unit tests cover configuration validation, directed-group initialization, deferred-receive forwarding, structured pipeline inputs, persistent loss output, and wrapper selection.

- `python -m pytest -q tests/unit_tests/cpu/test_cudagraph.py tests/unit_tests/cpu/test_trainer.py tests/unit_tests/cpu/test_config_manager.py tests/unit_tests/cpu/test_pipeline_parallel.py tests/unit_tests/cpu/test_integration_test_definitions.py`
- `PYTHONPATH=. python tests/integration_tests/run_tests.py <output> --test_suite features --test_name pp_looped_1f1b_cudagraph --ngpu 4` (10/10 steps)
- Deterministic four-GPU eager-versus-graph comparison with seed 42: bitwise-identical loss and `grad_norm` for all 10 steps

stack-info: PR: https://github.com/pytorch/torchtitan/pull/4435, branch: xmfan/stack/52

### CODE REVIEW

**Comment by xmfan** [comment]
Closing duplicate PR created by an incorrect ghstack resubmission; the original stack remains in #4429-#4436.
